File: app/src/main/java/remix/myplayer/ui/fragment/FolderFragment.java
package remix.myplayer.ui.fragment;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.loader.content.Loader;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;
import butterknife.BindView;
import java.util.List;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.Folder;
import remix.myplayer.misc.asynctask.WrappedAsyncTaskLoader;
import remix.myplayer.misc.interfaces.LoaderIds;
import remix.myplayer.misc.interfaces.OnItemClickListener;
import remix.myplayer.ui.activity.ChildHolderActivity;
import remix.myplayer.ui.adapter.FolderAdapter;
import remix.myplayer.util.Constants;
import remix.myplayer.util.MediaStoreUtil;

/**
 * Created by Remix on 2015/12/5.
 */

/**
 * 文件夹Fragment
 */
public class FolderFragment extends LibraryFragment<Folder, FolderAdapter> {

  @BindView(R.id.recyclerView)
  RecyclerView mRecyclerView;

  public static final String TAG = FolderFragment.class.getSimpleName();

  @Override
  protected int getLayoutID() {
    return R.layout.fragment_folder;
  }

  @Override
  protected void initAdapter() {
    mAdapter = new FolderAdapter(R.layout.item_folder_recycle, mChoice);
    mAdapter.setOnItemClickListener(new OnItemClickListener() {
      @Override
      public void onItemClick(View view, int position) {
        Folder folder = mAdapter.getDatas().get(position);
        String path = folder.getPath();
        if (getUserVisibleHint() && !TextUtils.isEmpty(path) &&
            !mChoice.click(position, folder)) {
          ChildHolderActivity.start(mContext, Constants.FOLDER, folder.getParentId(), path);
        }
      }

      @Override
      public void onItemLongClick(View view, int position) {
        Folder folder = mAdapter.getDatas().get(position);
        String path = mAdapter.getDatas().get(position).getPath();
        if (getUserVisibleHint() && !TextUtils.isEmpty(path)) {
          mChoice.longClick(position, folder);
        }
      }
    });
  }

  @Override
  protected void initView() {
    mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    mRecyclerView.setHasFixedSize(true);
    mRecyclerView.setItemAnimator(new DefaultItemAnimator());
    mRecyclerView.setAdapter(mAdapter);
  }

  @Override
  protected Loader<List<Folder>> getLoader() {
    return new AsyncFolderLoader(mContext);
  }

  @Override
  protected int getLoaderId() {
    return LoaderIds.FOLDER_FRAGMENT;
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    mPageName = TAG;
  }

  @Override
  public FolderAdapter getAdapter() {
    return mAdapter;
  }

  private static class AsyncFolderLoader extends WrappedAsyncTaskLoader<List<Folder>> {

    private AsyncFolderLoader(Context context) {
      super(context);
    }

    @Override
    public List<Folder> loadInBackground() {
      return MediaStoreUtil.getAllFolder();
    }
  }
}

File: app/src/main/java/remix/myplayer/ui/fragment/FolderFragment.kt
package remix.myplayer.ui.fragment

import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import androidx.loader.content.Loader
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.fragment_folder.*
import remix.myplayer.R
import remix.myplayer.bean.mp3.Folder
import remix.myplayer.misc.asynctask.WrappedAsyncTaskLoader
import remix.myplayer.misc.interfaces.LoaderIds
import remix.myplayer.misc.interfaces.OnItemClickListener
import remix.myplayer.ui.activity.ChildHolderActivity
import remix.myplayer.ui.adapter.FolderAdapter
import remix.myplayer.util.Constants
import remix.myplayer.util.ItemsSorter
import remix.myplayer.util.MediaStoreUtil.getAllFolder

/**
 * Created by Remix on 2015/12/5.
 */
/**
 * 文件夹Fragment
 */
class FolderFragment : LibraryFragment<Folder, FolderAdapter>() {

  override val layoutID = R.layout.fragment_folder

  override fun initAdapter() {
    adapter = FolderAdapter(R.layout.item_folder_recycle, multiChoice)
    adapter.onItemClickListener = object : OnItemClickListener {
      override fun onItemClick(view: View, position: Int) {
        val folder = adapter.dataList[position]
        val path = folder.path
        if (userVisibleHint && !TextUtils.isEmpty(path) && !multiChoice.click(position, folder)) {
          ChildHolderActivity.start(requireContext(), Constants.FOLDER, folder.path, path)
        }
      }

      override fun onItemLongClick(view: View, position: Int) {
        val folder = adapter.dataList[position]
        val path = adapter.dataList[position].path
        if (userVisibleHint && !TextUtils.isEmpty(path)) {
          multiChoice.longClick(position, folder)
        }
      }
    }
  }

  override fun initView() {
    recyclerView.layoutManager = LinearLayoutManager(context)
    recyclerView.setHasFixedSize(true)
    recyclerView.itemAnimator = DefaultItemAnimator()
    recyclerView.adapter = adapter
  }

  override fun loader(): Loader<List<Folder>> {
    return AsyncFolderLoader(requireContext())
  }

  override val loaderId: Int = LoaderIds.FRAGMENT_FOLDER

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    pageName = TAG
  }

  private class AsyncFolderLoader(context: Context?) : WrappedAsyncTaskLoader<List<Folder>>(context) {
    override fun loadInBackground(): List<Folder> {
      return getAllFolder().sortedBy { it.name }
    }
  }

  companion object {
    val TAG = FolderFragment::class.java.simpleName
  }
}